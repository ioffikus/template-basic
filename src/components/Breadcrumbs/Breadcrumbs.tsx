import React from 'react';
import { connect } from 'react-redux';
import { Breadcrumb } from 'antd';
import { Link } from 'react-router-dom';
import _ from 'lodash';
import queryString from 'query-string';
import getCategoriesPath from 'src/helpers/getCategoriesPath';

interface IOwnProps {
  for: 'product' | 'catalog';
}

interface IConnectedState {
  categoriesPath: Alicanto.Models.Category[];
}

type IProps = IOwnProps & IConnectedState;

interface IState {}

const mapStateToProps = (state: Store.IState, ownProps: IOwnProps) => ({
  categoriesPath:
    (ownProps.for === 'product' && state.product.categories) ||
    (ownProps.for === 'catalog' &&
      getCategoriesPath(
        state.categoriesMenu.content,
        _.get(state.catalogQuery.filterObj, 'categories__in[0]', null),
      )) ||
    [],
});

const WrappedBreadcrumbs = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class Breadcrumbs extends React.Component<IProps, IState> {
    render() {
      const openCategories = [...this.props.categoriesPath];
      const selectedCategory = openCategories.pop();
      if (!selectedCategory) {
        return null;
      }
      return (
        <div className="root">
          <style jsx>{`
            .root :global(.ant-breadcrumb a) {
              color: #b2bdcb;
              font-size: 14px;
              letter-spacing: 2px;
              font-weight: bold;
            }
            .root :global(.ant-breadcrumb-separator) {
              margin: 0 14px;
            }
            .root :global(.ant-breadcrumb > span:last-child) {
              color: #062f49;
              font-size: 2em;
              font-weight: bold;
            }
          `}</style>
          <Breadcrumb separator=" ">
            {openCategories.map(category => (
              <Breadcrumb.Item key={category.id}>
                <Link
                  to={{
                    pathname: '/catalog',
                    search: queryString.stringify({
                      filter: JSON.stringify({
                        categories__in: [category.id],
                      }),
                    }),
                  }}
                >
                  {category.name}
                </Link>
              </Breadcrumb.Item>
            ))}
            <Breadcrumb.Item key={selectedCategory.id}>{selectedCategory.name}</Breadcrumb.Item>
          </Breadcrumb>
        </div>
      );
    }
  },
);

export default WrappedBreadcrumbs;
