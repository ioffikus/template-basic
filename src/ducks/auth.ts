import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import { push } from 'connected-react-router';
import { GET, POST } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'AUTH';
const NS = `${REDUCER}__`;

export interface IAuthState extends Store.IReduxCommonState {
  isAuthenticated: boolean;
  isAuthLinkRequesting: boolean;
  url: string;
  data: object;
}

export const initialState: IAuthState = {
  isAuthenticated: false,
  isRequesting: false,
  isRequestError: false,
  isAuthLinkRequesting: false,
  url: null,
  data: null,
};

const reducer = createReducer<IAuthState>({}, initialState);

export const loginUserSuccess = createAction(`${NS}LOGIN_USER_SUCCESS`);
reducer.on(loginUserSuccess, state => ({
  ...state,
  isRequesting: false,
  isRequestError: false,
  isAuthenticated: true,
}));

const authLinkSuccess = createAction<{ url: string; data: object }>(`${NS}AUTH_LINK_SUCCESS`);
reducer.on(authLinkSuccess, (state, { url, data }) => ({
  ...state,
  url,
  data: { ...data },
  isAuthLinkRequesting: false,
}));

const loginUserFailure = createAction(`${NS}LOGIN_USER_FAILURE`);
reducer.on(loginUserFailure, state => ({
  ...state,
  isRequesting: false,
  isAuthenticated: false,
  isRequestError: true,
}));

const authLinkFailure = createAction(`${NS}AUTH_LINK_FAILURE`);
reducer.on(authLinkFailure, state => ({
  ...state,
  isAuthLinkRequesting: false,
  url: null,
  data: null,
}));

const loginUserRequest = createAction(`${NS}LOGIN_USER_REQUEST`);
reducer.on(loginUserRequest, state => ({
  ...state,
  isRequesting: true,
}));

const authLinkRequest = createAction(`${NS}AUTH_LINK_REQUEST`);
reducer.on(authLinkRequest, state => ({
  ...state,
  isAuthLinkRequesting: true,
  url: null,
  data: null,
}));

const reset = createAction(`${NS}RESET`);
reducer.on(reset, () => ({
  ...initialState,
}));

export const getAuthLink = (): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(authLinkRequest());
    GET(
      {
        url: `${EndpointsConfig.AUTH.AUTH_LINK}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        const { url, data } = resp.data as { url: string; data: object };
        // WIP does not use Alicanto.API.GET_AUTH_OAUTH_PROVIDER_URL_RESPONSE for develop
        dispatch(authLinkSuccess({ url, data }));
        resolve();
      })
      .catch(error => {
        dispatch(authLinkFailure());
        reject(error);
      });
  });
};

export const loginUser = (username: string, password: string): Store.IReduxThunkCb => (
  dispatch,
  getState,
  xShopHeaders,
) => {
  return new Promise((resolve, reject) => {
    dispatch(loginUserRequest());
    const data: Alicanto.API.POST_AUTH_LOGIN_BODY = {
      username,
      password,
    };
    POST(
      {
        data,
        url: `${EndpointsConfig.AUTH.LOGIN}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        dispatch(loginUserSuccess());
        if (process.env.BROWSER) {
          // TODO: move to external
          // redirect to home page
          dispatch(push('/'));
        }
        resolve();
      })
      .catch(error => {
        dispatch(loginUserFailure());
        reject(error);
      });
  });
};

export const logoutUser = (cb: () => void): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(loginUserRequest());
    GET(
      {
        url: `${EndpointsConfig.AUTH.LOGOUT}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        dispatch(loginUserSuccess());
        dispatch(reset());
        if (cb) {
          cb();
        }
        if (process.env.BROWSER) {
          // TODO: move to external
          // redirect to home page
          dispatch(push('/'));
        }
        resolve();
      })
      .catch(error => {
        dispatch(loginUserFailure());
        reject(error);
      });
  });
};

export default reducer;
