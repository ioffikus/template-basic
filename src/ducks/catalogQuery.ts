import { createAction, createReducer } from 'redux-act';
import queryString from 'query-string';
import _ from 'lodash';
import { safelyParseJSON } from 'src/helpers/safelyParseJSON';
import { IObject } from 'src/core/interfaces/IObject';

export const REDUCER = 'CATALOG_QUERY';
const NS = `${REDUCER}__`;

export interface ICatalogQueryState {
  search: string;
  page: string;
  pageSize: string;
  ordering: string;
  filterObj: IObject<Alicanto.API.CatalogFilter>;
}

export const initialState: ICatalogQueryState = {
  search: '',
  page: '1',
  pageSize: '12',
  ordering: '',
  filterObj: {},
};

const reducer = createReducer<ICatalogQueryState>({}, initialState);

const set = createAction<ICatalogQueryState>(`${NS}SET`);
reducer.on(set, (state, { search, page, pageSize, ordering, filterObj }) => ({
  ...state,
  search,
  page,
  pageSize,
  ordering,
  filterObj: { ...filterObj },
}));

export const getCatalogQuery = (): Store.IReduxThunkCb => (dispatch, getState) => {
  const router = getState().router;
  const query = queryString.parse(router.location.search);
  const search = _.get(query, 'search', initialState.search);
  const page = _.get(query, 'page', initialState.page);
  const pageSize = _.get(query, 'page_size', initialState.pageSize);
  const ordering = _.get(query, 'ordering', initialState.ordering);
  const filterStr = _.get(query, 'filter', null);
  const filterObj = safelyParseJSON(filterStr) || {};
  dispatch(set({ search, page, pageSize, ordering, filterObj }));
  return Promise.resolve();
};

export default reducer;
