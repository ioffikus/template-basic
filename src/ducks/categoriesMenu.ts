import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import { GET } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'CATEGORIES_MENU';
const NS = `${REDUCER}__`;

interface ICategoriesMenuResponse extends Alicanto.API.GET_CATALOG_GROUPS_SLUG_RESPONSE {
  content: Alicanto.Models.Category[];
}

export interface ICategoriesMenuState extends ICategoriesMenuResponse, Store.IReduxCommonState {}

export const initialState: ICategoriesMenuState = {
  slug: '',
  type: '',
  content: [],
  isRequesting: false,
  isRequestError: false,
};

const reducer = createReducer<ICategoriesMenuState>({}, initialState);

const callSuccess = createAction<ICategoriesMenuResponse>(`${NS}CALL_SUCCESS`);
reducer.on(callSuccess, (state, { slug, type, content }) => {
  return {
    ...state,
    slug,
    type,
    content: [...content],
    isRequestError: false,
    isRequesting: false,
  };
});

const callFailure = createAction(`${NS}CALL_FAILURE`);
reducer.on(callFailure, state => ({
  ...state,
  slug: '',
  type: '',
  content: [],
  isRequesting: false,
  isRequestError: true,
}));

const callRequest = createAction(`${NS}CALL_REQUEST`);
reducer.on(callRequest, state => ({
  ...state,
  isRequesting: true,
}));

export const getCategoriesMenu = (abortToken: string): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    GET(
      {
        url: EndpointsConfig.CATEGORIES_MENU,
      },
      xShopHeaders,
      abortToken,
    )
      .then((resp: AxiosResponse) => {
        const data = resp.data as ICategoriesMenuResponse;
        dispatch(callSuccess(data));
        resolve();
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

export default reducer;
