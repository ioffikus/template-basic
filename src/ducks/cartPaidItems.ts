import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import { ICardPaidItemsState } from 'src/ducks/cartPaidItems';
import _ from 'lodash';
import { GET } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'CART_PAID_ITEMS';
const NS = `${REDUCER}__`;

export interface ICardPaidItemsState extends Alicanto.API.GET_CART_PAID_ITEMS_RESPONSE, Store.IReduxCommonState {}

export const initialState: ICardPaidItemsState = {
  next: null,
  previous: null,
  count: 0,
  results: [],
  isRequestError: false,
  isRequesting: false,
};

const reducer = createReducer<ICardPaidItemsState>({}, initialState);

const callFailure = createAction(`${NS}CALL_FAILURE`);
reducer.on(callFailure, state => ({
  ...state,
  next: null,
  previous: null,
  count: 0,
  results: [],
  isRequestError: true,
  isRequesting: false,
}));

const callSuccess = createAction<Alicanto.API.GET_CART_PAID_ITEMS_RESPONSE>(`${NS}CALL_SUCCESS`);
reducer.on(callSuccess, (state, { next, previous, count, results }) => ({
  ...state,
  next,
  previous,
  count,
  results: [...results],
  isRequestError: false,
  isRequesting: false,
}));

const callRequest = createAction(`${NS}CALL_REQUEST`);
reducer.on(callRequest, state => ({
  ...state,
  isRequesting: true,
}));

interface ICartPaidItemsQuery {
  page?: string;
  page_size?: string;
}

const getCartPaidItemsRequestParams = ({ page, page_size }: ICartPaidItemsQuery) => {
  const params: ICartPaidItemsQuery = {
    page_size: '10',
  };
  if (!_.isNil(page)) {
    params.page = page;
  }
  if (!_.isNil(page_size)) {
    params.page_size = page_size;
  }
  return params;
};

export const getCartPaidItems = (page = '1'): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    GET(
      {
        url: `${EndpointsConfig.CART_PAID_ITEMS}`,
        params: getCartPaidItemsRequestParams({ page }),
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        const data = resp.data as Alicanto.API.GET_CART_PAID_ITEMS_RESPONSE;
        let results = data.results;
        if (page !== '1') {
          const oldResults = getState().cartPaidItems.results;
          results = [...oldResults, ...results];
        }
        dispatch(callSuccess({ ...data, results }));
        resolve();
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

export default reducer;
