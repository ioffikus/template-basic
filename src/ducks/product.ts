import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import { matchPath } from 'react-router';
import _ from 'lodash';
import { GET } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';
import { setMeta } from 'src/ducks/meta';

export const REDUCER = 'PRODUCTS';
const NS = `${REDUCER}__`;

export interface IProductState extends Alicanto.API.GET_CATALOG_UID_RESPONSE, Store.IReduxCommonState {}

export const initialState: IProductState = {
  uid: null,
  id: null,
  is_unlimited: null,
  schema_fields: {},
  categories: null,
  modifications: null,
  pictures: null,
  isRequesting: false,
  isRequestError: false,
};

const reducer = createReducer<IProductState>({}, initialState);

const callSuccess = createAction<Alicanto.API.GET_CATALOG_UID_RESPONSE>(`${NS}CALL_SUCCESS`);
reducer.on(callSuccess, (state, { id, uid, schema_fields, categories, modifications, pictures, is_unlimited }) => ({
  ...state,
  uid,
  id: _.isNil(id) ? null : id,
  is_unlimited: _.isNil(is_unlimited) ? null : is_unlimited,
  schema_fields: { ...schema_fields },
  categories: _.isNil(categories) ? null : [...categories],
  modifications: _.isNil(modifications) ? null : [...modifications],
  pictures: _.isNil(pictures) ? null : [...pictures],
  isRequesting: false,
  isRequestError: false,
}));

const callFailure = createAction(`${NS}CALL_FAILURE`);
reducer.on(callFailure, state => ({
  ...state,
  uid: null,
  id: null,
  is_unlimited: null,
  schema_fields: {},
  categories: null,
  modifications: null,
  pictures: null,
  isRequestError: true,
}));

const callRequest = createAction(`${NS}CALL_REQUEST`);
reducer.on(callRequest, state => ({
  ...state,
  isRequesting: true,
}));

export const getProduct = (abortToken: string): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    const router = getState().router;
    const match = matchPath(router.location.pathname, { path: '/product/:uid' });
    const uid = _.get(match, 'params.uid', null);
    GET(
      {
        url: `${EndpointsConfig.PRODUCT}${uid}`,
      },
      xShopHeaders,
      abortToken,
    )
      .then((resp: AxiosResponse) => {
        const data = resp.data as Alicanto.API.GET_CATALOG_UID_RESPONSE;
        dispatch(callSuccess(data));
        resolve();
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  }).then(() => {
    const schemaFields = getState().product.schema_fields;
    const name = _.get(schemaFields, 'name.value', '');
    const description = _.get(schemaFields, 'description.value', '');
    dispatch(setMeta({ description, title: name, keywords: `${name} ${description}` }));
  });
};

export default reducer;
