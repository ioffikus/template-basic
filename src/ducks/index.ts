// instead this, we have router in store via connected-react-router
// import { routerReducer as routing, RouterState } from 'react-router-redux';
import { combineReducers } from 'redux';
import auth from './auth';
import cart from './cart';
import cartPaidItems from './cartPaidItems';
import catalogQuery from './catalogQuery';
import catalog from './catalog';
import categories from './categories';
import categoriesMenu from './categoriesMenu';
import filter from './filter';
import locale from './locale';
import meta from './meta';
import page from './page';
import product from './product';
import progress from './progress';
import popularItems from './popularItems';
import recommendItems from './recommendItems';
import routes from './routes';
import ui from './ui';
import user from './user';
import userDelivery from './userDelivery';
import userOrders from './userOrders';

export default combineReducers<Store.IState>({
  ui,
  auth,
  user,
  categories,
  categoriesMenu,
  catalogQuery,
  catalog,
  product,
  progress,
  routes,
  meta,
  filter,
  cart,
  cartPaidItems,
  userDelivery,
  userOrders,
  popularItems,
  recommendItems,
  page,
  locale,
});
