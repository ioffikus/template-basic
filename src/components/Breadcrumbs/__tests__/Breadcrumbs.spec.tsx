import React from 'react';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import uiStateMock from 'src/store/__mocks__/uiStore';
import productStateMock from 'src/store/__mocks__/productStore';
import catalogQueryStateMock from 'src/store/__mocks__/catalogQueryStore';
import categoriesMenuStateMock from 'src/store/__mocks__/categoriesMenuStore';
import { initialState as productInitialState } from 'src/ducks/product';
import { initialState as catalogQueryInitialState } from 'src/ducks/catalogQuery';
import { initialState as categoriesMenuInitialState } from 'src/ducks/categoriesMenu';
import Breadcrumbs from '../Breadcrumbs';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe(null, () => {
  it('render() with product initialiItems ', () => {
    const store = mockStore({ ui: uiStateMock, product: productInitialState });
    const wrapper = shallow(<Breadcrumbs for="product" />, { context: { store } });
    expect(wrapper.dive()).toMatchSnapshot();
  });
  it('render() with product items ', () => {
    const store = mockStore({ ui: uiStateMock, product: productStateMock });
    const wrapper = shallow(<Breadcrumbs for="product" />, { context: { store } });
    expect(wrapper.dive()).toMatchSnapshot();
  });
  it('render() with categories initialiItems ', () => {
    const store = mockStore({
      ui: uiStateMock,
      categoriesMenu: categoriesMenuInitialState,
      catalogQuery: catalogQueryInitialState,
    });
    const wrapper = shallow(<Breadcrumbs for="catalog" />, { context: { store } });
    expect(wrapper.dive()).toMatchSnapshot();
  });
  it('render() with categories', () => {
    const store = mockStore({
      ui: uiStateMock,
      categoriesMenu: categoriesMenuStateMock,
      catalogQuery: catalogQueryStateMock,
    });
    const wrapper = shallow(<Breadcrumbs for="catalog" />, { context: { store } });
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
