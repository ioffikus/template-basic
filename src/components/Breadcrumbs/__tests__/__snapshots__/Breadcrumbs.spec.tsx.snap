// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render() with categories 1`] = `
<div
  className="jsx-2632493999 root"
>
  <JSXStyle
    css=".root.jsx-2632493999 .ant-breadcrumb a{color:#b2bdcb;font-size:14px;-webkit-letter-spacing:2px;-moz-letter-spacing:2px;-ms-letter-spacing:2px;letter-spacing:2px;font-weight:bold;}.root.jsx-2632493999 .ant-breadcrumb-separator{margin:0 14px;}.root.jsx-2632493999 .ant-breadcrumb > span:last-child{color:#062f49;font-size:2em;font-weight:bold;}"
    styleId="2632493999"
  />
  <Breadcrumb
    prefixCls="ant-breadcrumb"
    separator=" "
  >
    <BreadcrumbItem
      key="35"
      prefixCls="ant-breadcrumb"
      separator="/"
    >
      <Link
        replace={false}
        to={
          Object {
            "pathname": "/catalog",
            "search": "filter=%7B%22categories__in%22%3A%5B35%5D%7D",
          }
        }
      >
        books
      </Link>
    </BreadcrumbItem>
    <BreadcrumbItem
      key="37"
      prefixCls="ant-breadcrumb"
      separator="/"
    >
      eBooks
    </BreadcrumbItem>
  </Breadcrumb>
</div>
`;

exports[`render() with categories initialiItems  1`] = `""`;

exports[`render() with product initialiItems  1`] = `""`;

exports[`render() with product items  1`] = `
<div
  className="jsx-2632493999 root"
>
  <JSXStyle
    css=".root.jsx-2632493999 .ant-breadcrumb a{color:#b2bdcb;font-size:14px;-webkit-letter-spacing:2px;-moz-letter-spacing:2px;-ms-letter-spacing:2px;letter-spacing:2px;font-weight:bold;}.root.jsx-2632493999 .ant-breadcrumb-separator{margin:0 14px;}.root.jsx-2632493999 .ant-breadcrumb > span:last-child{color:#062f49;font-size:2em;font-weight:bold;}"
    styleId="2632493999"
  />
  <Breadcrumb
    prefixCls="ant-breadcrumb"
    separator=" "
  >
    <BreadcrumbItem
      key="115"
      prefixCls="ant-breadcrumb"
      separator="/"
    >
      <Link
        replace={false}
        to={
          Object {
            "pathname": "/catalog",
            "search": "filter=%7B%22categories__in%22%3A%5B115%5D%7D",
          }
        }
      >
        books
      </Link>
    </BreadcrumbItem>
    <BreadcrumbItem
      key="117"
      prefixCls="ant-breadcrumb"
      separator="/"
    >
      <Link
        replace={false}
        to={
          Object {
            "pathname": "/catalog",
            "search": "filter=%7B%22categories__in%22%3A%5B117%5D%7D",
          }
        }
      >
        eBooks
      </Link>
    </BreadcrumbItem>
    <BreadcrumbItem
      key="118"
      prefixCls="ant-breadcrumb"
      separator="/"
    >
      <Link
        replace={false}
        to={
          Object {
            "pathname": "/catalog",
            "search": "filter=%7B%22categories__in%22%3A%5B118%5D%7D",
          }
        }
      >
        PDF
      </Link>
    </BreadcrumbItem>
    <BreadcrumbItem
      key="81"
      prefixCls="ant-breadcrumb"
      separator="/"
    >
      <Link
        replace={false}
        to={
          Object {
            "pathname": "/catalog",
            "search": "filter=%7B%22categories__in%22%3A%5B81%5D%7D",
          }
        }
      >
        Speculative fiction
      </Link>
    </BreadcrumbItem>
    <BreadcrumbItem
      key="82"
      prefixCls="ant-breadcrumb"
      separator="/"
    >
      Science fiction
    </BreadcrumbItem>
  </Breadcrumb>
</div>
`;
